### PR TITLE
Gui: fix macOS trackpad navigation (pinch zoom, scroll pan, Shift orbit)

### LINKED ISSUE #5938: Navigation: Add Trackpad gestures support for macOS
Issue imported from https://tracker.freecad.org/view.php?id=3995

* **Reporter:** @@robot-army
* **Date submitted:** 5/31/2019
* **FreeCAD version:** 0.18
* **Category:** Bug
* **Status:** new
* **Tags:** Bountysource Potential, gesture, macOS, trackpad

# Original report text

Hi all, 

My trackpad and FreeCAD don't get along :( 

None of the two-finger things (pan, pinch to zoom) behave as I would expect, and it makes it almost impossible to use FreeCAD on my normal platform. 

Forum posts related: 
https://forum.freecadweb.org/viewtopic.php?f=3&t=34351
('Gesture mode' as is suggested doesn't work at all well for me)

https://forum.freecadweb.org/viewtopic.php?f=3&t=33282&start=10
(If lack of mac is the root cause, I'd be happy to loan one! Good luck DeepSOIC/all)

I think this may also be related to #2437 but that's listed as a 'minor feature request' on 0.15. This is a showstopper for me using FreeCAD. 

I've captured two videos with the default settings, and I'm not seeing what I would consider 'expected behaviour' with my trackpad, and none of the mouse models make it much more useful. 

You can see what I see here with only very minor finger movements in each case. 
https://www.youtube.com/watch?v=iA9-QKA6iQ8
https://www.youtube.com/watch?v=6hmtKJJUAKM

It 'should' behave (I suppose?) as described here: https://developer.apple.com/design/human-interface-guidelines/macos/user-interaction/mouse-and-trackpad/

Maybe setting some zoom multiplier would help it not be so wild, but the gestures are being interpreted incorrectly, as well as being too fast. 

The 'Gesture' model looks the most correct on the mouseover, but it seems that the two finger gestures are being picked up as scrolling (or something?) rather than the gestures they are. 

Thanks for all your hard work crew, it's come a long way!!

# Additional information

OS: macOS Sierra (10.12)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.18.16117 (Git)
Build type: Release
Branch: (HEAD detached at 0.18.2)
Hash: dbb4cc6415bac848a294f03b80f65e888d531742
Python version: 3.6.7
Qt version: 5.6.2
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: English/UnitedKingdom (en_GB)

# Steps to reproduce

Run 0.18 on a Macbook with the mouse model set to 'Gesture' mode and try to do two-finger gestures. 

# Other bug information

* **Priority:** normal
* **Severity:** major
* **Category:** Bug
* **Updated:** 2/6/2021

# Discussion from Mantis ticket

### Comment by @luzpaz 2019-06-17 06:23
robotarmy please participate in https://forum.freecadweb.org/viewtopic.php?f=9&t=36387 
DeepSOIC has updated the gesture navigation code in 0.19_pre but mostly for windows. We need MacOS X testers that can give him feedback (because he doesn't own a mac).

---

### Comment by krite 2019-10-25 15:07
This issue is exactly as bad in Ubuntu on a Lenovo T480s, totally unusable.
The behaviour is also identical in the newest 19_pre from Sep 7.

Further analysis actually show that the scroll upp/down on the trackpad behaves as pressing pgup/pgdown-keys. Thus not zooming like a mouse at all, which is totally unusable of course.

The two-finger up/down-scroll is thus not usable (large hops), and the pinch-zoom is actually intercepted incorrectly and interfer with the two-finger action, and does not work at all.

---

### Comment by @luzpaz 2020-09-16 13:07
DeepSOIC were you working on this ?

---

### Comment by @DeepSOIC 2020-09-16 17:15
i am not. Fixing gigantic zoom sensitivity ( https://forum.freecadweb.org/viewtopic.php?f=9&t=49680 ) should have made things a lot better, but i don't have a macbook to try (and thus i can't work on it). chrisb ?

---

### Comment by @DeepSOIC 2020-09-16 17:17
As for the linux part, I can try. I don't have linux, but i can install one in theory.

### PR DESCRIPTION
**Pull request by pavliha**
- [x] This PR is not unverified AI output, I take responsibility for it, and all communication from my side in this PR is done by me personally.

## Issues

Closes #5938. Related: #5640 (closed) and #29447. #29447 covers the pinch half of the same problem; see "Relationship to #29447" below.

## Problem

On macOS, three trackpad inputs do nothing in the 3D view. Pinch-to-zoom is ignored under the default CAD navigation style, two-finger rotate never works, and horizontal two-finger scroll is discarded.

## Root cause

macOS delivers trackpad pinch and rotate as `QNativeGestureEvent`s. Qt's `QWidgetWindow::handleGestureEvent` sends those to `QApplication::widgetAt(pos)`, the deepest widget under the cursor, and gesture events do not propagate to parents.

For the QGraphicsView-based 3D view, the deepest widget is the GL viewport. Every one of FreeCAD's 3D-view event filters is installed on the view widget instead, one level up. So the gestures have been arriving below every listener and dying there.

(Verified against the Qt 6.8 sources `qnsview_gestures.mm` and `qwidgetwindow.cpp`, and empirically with event tracing on Qt 6.8.3.)

## Changes

- Install `ViewerEventFilter` on the viewport as well, and translate native zoom/rotate gestures into `SoGesturePinchEvent`s.
- Handle pinch in the `NavigationStyle` base class (`processPinchEvent`), so every navigation style gets it instead of only the two gesture styles.
- Carry `QWheelEvent::pixelDelta()` through `SoMouseWheelEvent`. On devices that report pixel-precise deltas (touchpads), two-finger scroll pans, Shift+scroll orbits, and Ctrl+scroll (Cmd on macOS) zooms, matching Fusion 360 and Onshape. Wheel mice report no pixel delta and keep the current zoom behavior on every platform. The scheme sits behind a new preference (`View/TouchpadScrollPans`, checkbox in Preferences → Navigation) that defaults to on for macOS and off elsewhere.
- Stop grabbing Qt pan/pinch gestures on macOS. Qt's recognizer synthesizes a second pinch from the same touch stream, always with `rotationAngle == 0` because macOS never sends the native rotate stream Qt sources it from, and that second pinch would double the zoom.
- Restrict the horizontal-wheel filter (bug #0000607) to non-precise devices, so trackpads keep both scroll axes.
- Fix a null-pointer dereference in the swipe branch of `GesturesDevice::translateEvent`, which built a `SoGesturePanEvent` from a null `QPanGesture*`.

## Relationship to #29447

#29447 translates `QNativeGestureEvent` inside `GesturesDevice::translateEvent()` and also fixes the same swipe-branch null pointer. Two differences here:

1. As @chennes noted on that PR, only GestureNavigationStyle and MayaGestureNavigationStyle consume `SoGesturePinchEvent`, so translating the gesture is a partial fix: under the default CAD style the pinch is still dropped. This PR handles pinch in the `NavigationStyle` base class, so it reaches CAD, Blender, Touchpad and the rest too.
2. In my testing on Qt 6.8.3, `QNativeGestureEvent`s never reach the Quarter `EventFilter` at all. They are delivered to the GL viewport and stop there, which is why this PR hooks the viewport. I'd welcome the author of #29447 checking whether their branch still receives the events on current Qt.

Happy to coordinate so the swipe-branch fix lands only once.

## Behavioral note for touchpad users

On macOS this changes the default two-finger-scroll action on touchpads from zoom to pan, with pinch and Cmd+scroll doing the zooming, matching other CAD tools on that platform. On other platforms nothing changes by default: the same scheme is available behind the new preference, but the default stays zoom. The runtime gate is device capability plus the preference, wheel mice are never affected.

## Testing

https://github.com/user-attachments/assets/8ff2f959-3b2f-469e-a75a-a57621e914d4

https://github.com/user-attachments/assets/d4489917-837f-4706-be56-e23a731abbea

https://github.com/user-attachments/assets/6e8eeadf-f26d-412d-bdff-fe47c41c1c15

https://github.com/user-attachments/assets/69fc4a38-3431-4e38-8f82-2a50968bca63

https://github.com/user-attachments/assets/77d796a6-9b10-4b6b-88c9-3f50d1ffd211

macOS 15 (arm64), Qt 6.8.3, pixi/conda build:

- Manual: pinch zoom at cursor, two-finger pan (both axes), Shift-orbit, Cmd-zoom, under both CAD and Gesture navigation styles.
- Automated: synthesized pixel-precise scroll events via CGEventPost (with real modifier-key state) and asserted camera semantics through the Python API between batches. Pan moves the focal point purely laterally along the camera-up axis, with orientation and height untouched. Shift-orbit changes orientation, and camera elevation follows finger direction. Cmd-zoom changes only the view height. The assertions distinguish real motion from the orthographic focal-distance normalization that preserves the focal point.
- Screen recordings are embedded above.
- Unit tests: `tests/src/Gui/SoMouseWheelEventTest.cpp` covers the new `SoMouseWheelEvent` pixel-delta contract (`isPrecise()` semantics, including the horizontal-only case the old wheel filter used to discard). The navigation routing needs a live viewer with a GL context, so it is covered by the automated camera-semantics assertions above rather than unit tests.

Non-macOS platforms: no default behavior change. Wheel mice report no pixel delta so the existing zoom path runs, and the touchpad scheme is opt-in via the preference. The native gesture handler runs everywhere but only receives events where the platform produces them (macOS, Wayland). I could only test on macOS hardware, so testers on Linux and Windows touchpads are welcome.

## AI disclosure

Per the project's AI policy: this change was developed with AI assistance (Claude Opus 5, `claude-opus-5`, disclosed in the commit trailer). I directed the work, built and ran it locally, and tested the resulting behavior on my machine. I take responsibility for the contribution.

### CODE REVIEW

File: tests/src/Gui/SoMouseWheelEventTest.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+TEST(SoMouseWheelEvent, wheelDeltaDoesNotImplyPrecision)
+{
+    SoMouseWheelEvent event;
+
+    event.setDelta(120);
+    event.setPixelDelta(SbVec2f(0.0F, 0.0F));
+
+    EXPECT_FALSE(event.isPrecise());
+}
```
Comment (Lgt2x) [inline tests/src/Gui/SoMouseWheelEventTest.cpp:34]: these tests make no sense, why would you need 7 test cases to test a function that returns false if the vector is null?

File: src/Gui/Inventor/SoMouseWheelEvent.h [hunk 2]
```diff
@@ -59,8 +60,21 @@ class GuiExport SoMouseWheelEvent: public SoEvent
...
+        return pixelDelta;
+    }
+    void setPixelDelta(const SbVec2f& pixelDelta)
+    {
+        this->pixelDelta = pixelDelta;
+    }
+    bool isPrecise() const
+    {
+        return pixelDelta != SbVec2f(0.0F, 0.0F);
+    }
```
Comment (Lgt2x) [inline src/Gui/Inventor/SoMouseWheelEvent.h:74]: these functions are not connected to the rest of the class, I would expect pixel delta to be computed by this class starting using the pixel ratio

File: src/Gui/View3DInventorViewer.cpp [hunk 3]
```diff
@@ -798,22 +799,83 @@ class Gui::ViewerEventFilter: public QObject
     QPoint pressPosition;
     View3DInventorViewer* currentViewer = nullptr;
 
+#ifdef Q_OS_MACOS
```
Comment (Lgt2x) [inline src/Gui/View3DInventorViewer.cpp]: other systems can't use this code? I don't see anything system-specific

**Comment by Lgt2x** [comment]
@chennes do you want this? otherwise I'm fine with dropping the PR as it is low-quality LLM code

File: tests/src/Gui/SoMouseWheelEventTest.cpp (hunk 1 above)
Comment (pavliha) [reply tests/src/Gui/SoMouseWheelEventTest.cpp:34]: Yeah True. Thanks I removed them

File: src/Gui/View3DInventorViewer.cpp (hunk 3 above)
Comment (pavliha) [reply src/Gui/View3DInventorViewer.cpp]: I did try to make it mac os only so I don't affect other OS by chance as I don't have windows machine to test it there. But yeah i'll remove this guard

**Comment by pavliha** [comment]
I think I can do more testing and share more videos how it works I really like it. I can finally use FreeCAD on my mac with this fix

File: src/Gui/Inventor/SoMouseWheelEvent.h (hunk 2 above)
Comment (pavliha) [reply src/Gui/Inventor/SoMouseWheelEvent.h:74]: yeah agreed, I moved it into the class

**Comment by chennes** [comment]
The big problem here, @pavliha, is that your write up looks like a clear violation of our "no AI-written communication" policy. Please feel free to re-submit with a human-authored PR write up.
